server/src/main/java/org/candlepin/resource/OwnerResource.java

```diff
@@ -264,6 +291,75 @@ private Owner findOwner(String key) {
         return owner;
     }
 
+    /**
+     * Returns the owner object that is identified by the given id or key, if it is found in the system.
+     * A search with the id is performed first; if not found, a search with the id is performed.
+     * If it is also not found by id, a NotFoundException is thrown.
+     *
+     * @param id the id that is associated with the owner we are searching for.
+     *
+     * @param key the key that is associated with the owner we are searching for.
+     *
+     * @return the owner that was found in the system based on the given id or key.
+     *
+     * @throws NotFoundException
+     *  if the owner with the given id and key was not found in the system.
+     */
+    private Owner findOwnerByIdOrKey(String id, String key) {
+        Owner owner = null;
+
+        if (id != null) {
+            owner = this.ownerCurator.find(id);
+        }
+
+        if (owner == null && key != null) {
```

Comment:
We need to be careful with these kinds of lookups. Since this will be initiated by client-input, we're looking to pick one or the other according to which has higher accuracy. In this case, use ID if it's present, otherwise use the key. This leaves us with a lookup like:

```
if (id != null) {
   // by-id lookup
   if (owner == null) {
      // error: could not find owner by id
   }
}
else if (key != null) {
   ... // key lookup
   if (owner == null) {
      // error: could not find owner by key
}
else {
   // error: no identifiers provided
}
```

Since that's basically just wrapped versions of the other lookups, we can avoid some extra translation bits and whatnot by just calling out to each:

```
if (id != null) {
   return this.findOwnerById(id);
}
else if (key != null) {
   return this.findOwnerByKey(key);
}
else {
   // error: no identifying info
}
```

Reply:
I wrote it this way because even though the id might not be null and we do a lookup with it, the lookup might fail, in which case we fall back to the key. If we use an 'else if', then the fallback to the key will never run. Or is it that we don't care about it?

Reply:
The else-if will use the key only if the ID is not present, which is what we want. We want to use the most specific identifier provided, not a mixture of them. If we're provided an owner object that has both, but they don't align, then that's malformed data -- we can't be certain which the client intended to use.